Datasets:

- [120 years of Olympic history: athletes and results](https://www.kaggle.com/datasets/heesoo37/120-years-of-olympic-history-athletes-and-results/code)
- 

## 1. Download the Dataset

Import the required libraries

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# to draw pictures in jupyter notebook
%matplotlib inline

Upload the dataset

In [7]:
athletes_df = pd.read_csv("../datasets/athlete_events.csv")
countries_df = pd.read_csv("../datasets/noc_regions.csv")

## 2. Data Preparation

In [11]:
athletes_df.head()

ID                      Name Sex   Age  Height  Weight            Team  \
0   1                 A Dijiang   M  24.0   180.0    80.0           China   
1   2                  A Lamusi   M  23.0   170.0    60.0           China   
2   3       Gunnar Nielsen Aaby   M  24.0     NaN     NaN         Denmark   
3   4      Edgar Lindenau Aabye   M  34.0     NaN     NaN  Denmark/Sweden   
4   5  Christine Jacoba Aaftink   F  21.0   185.0    82.0     Netherlands   

   NOC        Games  Year  Season       City          Sport  \
0  CHN  1992 Summer  1992  Summer  Barcelona     Basketball   
1  CHN  2012 Summer  2012  Summer     London           Judo   
2  DEN  1920 Summer  1920  Summer  Antwerpen       Football   
3  DEN  1900 Summer  1900  Summer      Paris     Tug-Of-War   
4  NED  1988 Winter  1988  Winter    Calgary  Speed Skating   

                              Event Medal  
0       Basketball Men's Basketball   NaN  
1      Judo Men's Extra-Lightweight   NaN  
2           Football Men's Football   NaN  
3       Tug-Of-War Men's Tug-Of-War  Gold  
4  Speed Skating Women's 500 metres   NaN

### Start with athletes_df

Basic info of pd.DataFrame

In [23]:
athletes_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 271116 entries, 0 to 271115
Data columns (total 15 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   ID      271116 non-null  int64  
 1   Name    271116 non-null  object 
 2   Sex     271116 non-null  object 
 3   Age     261642 non-null  float64
 4   Height  210945 non-null  float64
 5   Weight  208241 non-null  float64
 6   Team    271116 non-null  object 
 7   NOC     271116 non-null  object 
 8   Games   271116 non-null  object 
 9   Year    271116 non-null  int64  
 10  Season  271116 non-null  object 
 11  City    271116 non-null  object 
 12  Sport   271116 non-null  object 
 13  Event   271116 non-null  object 
 14  Medal   39783 non-null   object 
dtypes: float64(3), int64(2), object(10)
memory usage: 31.0+ MB


Checking for any missing values

In [21]:
athletes_df.isna().mean()

ID        0.000000
Name      0.000000
Sex       0.000000
Age       0.034944
Height    0.221938
Weight    0.231912
Team      0.000000
NOC       0.000000
Games     0.000000
Year      0.000000
Season    0.000000
City      0.000000
Sport     0.000000
Event     0.000000
Medal     0.853262
dtype: float64

Checking for the total number of rows & columns

In [22]:
athletes_df.shape

(271116, 15)

`Name` column contains duplicate entires. I am going to count the number of athletes in the dataframe

In [25]:
athletes_df["Name"].nunique()

134732

Kind of sports are represented in the dataframe

In [43]:
athletes_df["Sport"].nunique()

17

Check for the duplicated rows and drop them

In [36]:
athletes_df.duplicated().sum()

0

In [32]:
athletes_df.drop_duplicates(['ID', 'Name', 'Games', 'Sport', 'Event', 'Medal'], inplace=True)

In [35]:
athletes_df.shape

(269718, 15)

In [37]:
athletes_df.duplicated().sum()

0

We will work with the Winter Olympic Games. Let's choose "Winter" in the `Season`column

In [38]:
athletes_df = athletes_df[athletes_df["Season"] == "Winter"]

We can now make some conclusions regarding this dataframe
- It contains 271116 rows and 15 columns.
- There are 134732 athletes
- It contains 1385 duplicated values
- There are 66 kind of sports

I've created a new dataframe only with the Winter Olympic Games and without duplicates

### Take a look on NOC regions dataset

In [44]:
countries_df.head()

NOC       region                 notes
0  AFG  Afghanistan                   NaN
1  AHO      Curacao  Netherlands Antilles
2  ALB      Albania                   NaN
3  ALG      Algeria                   NaN
4  AND      Andorra                   NaN

In [45]:
countries_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 230 entries, 0 to 229
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   NOC     230 non-null    object
 1   region  227 non-null    object
 2   notes   21 non-null     object
dtypes: object(3)
memory usage: 5.5+ KB


### Let's merge two DataFrames by `NOC` column

In [46]:
merged_df = athletes_df.merge(countries_df, on="NOC", how="left")
merged_df.head()

ID                      Name Sex   Age  Height  Weight         Team  NOC  \
0   5  Christine Jacoba Aaftink   F  21.0   185.0    82.0  Netherlands  NED   
1   5  Christine Jacoba Aaftink   F  21.0   185.0    82.0  Netherlands  NED   
2   5  Christine Jacoba Aaftink   F  25.0   185.0    82.0  Netherlands  NED   
3   5  Christine Jacoba Aaftink   F  25.0   185.0    82.0  Netherlands  NED   
4   5  Christine Jacoba Aaftink   F  27.0   185.0    82.0  Netherlands  NED   

         Games  Year  Season         City          Sport  \
0  1988 Winter  1988  Winter      Calgary  Speed Skating   
1  1988 Winter  1988  Winter      Calgary  Speed Skating   
2  1992 Winter  1992  Winter  Albertville  Speed Skating   
3  1992 Winter  1992  Winter  Albertville  Speed Skating   
4  1994 Winter  1994  Winter  Lillehammer  Speed Skating   

                                Event Medal       region notes  
0    Speed Skating Women's 500 metres   NaN  Netherlands   NaN  
1  Speed Skating Women's 1,000 metres   NaN  Netherlands   NaN  
2    Speed Skating Women's 500 metres   NaN  Netherlands   NaN  
3  Speed Skating Women's 1,000 metres   NaN  Netherlands   NaN  
4    Speed Skating Women's 500 metres   NaN  Netherlands   NaN

### Datasets with continents, countries and theid codes

The dataset with country codes is paid for. That's why I have successfully scraped and saved it as `Country Codes.csv`

In [67]:
x = pd.read_html("https://www.worlddata.info/countrycodes.php")[0]

In [70]:
x.to_csv("../datasets/Country Codes.csv", index=False, encoding="utf-8")

In [71]:
del x

Read them

In [73]:
iso_to_noc = pd.read_csv("../datasets/Country Codes.csv")
continents_df = pd.read_csv("../datasets/country-and-continent-codes-list-csv_csv.csv")

In [75]:
iso_to_noc.head()

Country ISO 3166-1 alpha2 ISO 3166-1 alpha3  ISO 3166-1 numeric  \
0     Afghanistan                AF               AFG                   4   
1   Åland Islands                AX               ALA                 248   
2         Albania                AL               ALB                   8   
3         Algeria                DZ               DZA                  12   
4  American Samoa                AS               ASM                  16   

   IOC Fips 10 License Plate Domain  
0  AFG      AF           AFG    .af  
1  NaN     NaN            AX    .ax  
2  ALB      AL            AL    .al  
3  ALG      AG            DZ    .dz  
4  ASA      AQ           USA    .as

In [74]:
continents_df.head()

Continent_Name Continent_Code                                  Country_Name  \
0           Asia             AS              Afghanistan, Islamic Republic of   
1         Europe             EU                          Albania, Republic of   
2     Antarctica             AN  Antarctica (the territory South of 60 deg S)   
3         Africa             AF      Algeria, People's Democratic Republic of   
4        Oceania             OC                                American Samoa   

  Two_Letter_Country_Code Three_Letter_Country_Code  Country_Number  
0                      AF                       AFG             4.0  
1                      AL                       ALB             8.0  
2                      AQ                       ATA            10.0  
3                      DZ                       DZA            12.0  
4                      AS                       ASM            16.0

According to the `iso_to_noc` dataframe and for convenient work, I am going to rename some columns in the dataframes

In [80]:
continents_df.rename(columns={"Two_Letter_Country_Code": "alpha 2",
                             "Three_Letter_Country_Code": "alpha 3"},
                    inplace=True)

iso_to_noc.rename(columns={"ISO 3166-1 alpha2": "alpha 2",
                           "ISO 3166-1 alpha3": "alpha 3",
                           "ISO 3166-1 numeric": "numeric"},
                 inplace=True)

Using the `alpha 3` parameter, merge these dataframes

In [82]:
country_codes_df = iso_to_noc.merge(continents_df, on="alpha 3", how="left")
country_codes_df.head()

Country alpha 2_x alpha 3  numeric  IOC Fips 10 License Plate  \
0     Afghanistan        AF     AFG        4  AFG      AF           AFG   
1   Åland Islands        AX     ALA      248  NaN     NaN            AX   
2         Albania        AL     ALB        8  ALB      AL            AL   
3         Algeria        DZ     DZA       12  ALG      AG            DZ   
4  American Samoa        AS     ASM       16  ASA      AQ           USA   

  Domain Continent_Name Continent_Code  \
0    .af           Asia             AS   
1    .ax         Europe             EU   
2    .al         Europe             EU   
3    .dz         Africa             AF   
4    .as        Oceania             OC   

                               Country_Name alpha 2_y  Country_Number  
0          Afghanistan, Islamic Republic of        AF             4.0  
1                             Åland Islands        AX           248.0  
2                      Albania, Republic of        AL             8.0  
3  Algeria, People's Democratic Republic of        DZ            12.0  
4                            American Samoa        AS            16.0

In [86]:
country_codes_df.rename(columns={"IOC": "NOC"}, inplace=True)

### Final dataframe

Just merge `countries_df` and `country_codes_df`, using `NOC` parameter

In [90]:
df = merged_df.merge(country_codes_df[["NOC", "alpha 3", "Continent_Name"]], on="NOC", how="left")
df.head()

ID                      Name Sex   Age  Height  Weight         Team  NOC  \
0   5  Christine Jacoba Aaftink   F  21.0   185.0    82.0  Netherlands  NED   
1   5  Christine Jacoba Aaftink   F  21.0   185.0    82.0  Netherlands  NED   
2   5  Christine Jacoba Aaftink   F  25.0   185.0    82.0  Netherlands  NED   
3   5  Christine Jacoba Aaftink   F  25.0   185.0    82.0  Netherlands  NED   
4   5  Christine Jacoba Aaftink   F  27.0   185.0    82.0  Netherlands  NED   

         Games  Year  Season         City          Sport  \
0  1988 Winter  1988  Winter      Calgary  Speed Skating   
1  1988 Winter  1988  Winter      Calgary  Speed Skating   
2  1992 Winter  1992  Winter  Albertville  Speed Skating   
3  1992 Winter  1992  Winter  Albertville  Speed Skating   
4  1994 Winter  1994  Winter  Lillehammer  Speed Skating   

                                Event Medal       region notes alpha 3  \
0    Speed Skating Women's 500 metres   NaN  Netherlands   NaN     NLD   
1  Speed Skating Women's 1,000 metres   NaN  Netherlands   NaN     NLD   
2    Speed Skating Women's 500 metres   NaN  Netherlands   NaN     NLD   
3  Speed Skating Women's 1,000 metres   NaN  Netherlands   NaN     NLD   
4    Speed Skating Women's 500 metres   NaN  Netherlands   NaN     NLD   

  Continent_Name  
0         Europe  
1         Europe  
2         Europe  
3         Europe  
4         Europe